### PR TITLE
Adding sign up page and resovle #39

### PR DESCRIPTION
**Pull request by zhengye1**
Fews things I changed 
Back end:
- Modify ```WebSecurityConfig.java``` to add ```/api/signup``` to ```ignoringAntMatchers```, so csrf will not checking the sign up url
-  Since it need to check the username existing or not,  I get rid the ``` @PreAuthorize("hasRole('USER')")``` in ```findByUsername```, and also change the ```UserServiceTest.java``` file to pass the mvn test
- Add the custom exception handler to handle the conflict exception part, so it return meaningful message to the user in the front end 
Reference for this part [Spring Boot REST web service – Part 3 – Exception Handling and Validation using @ControllerAdvice, @Valid and Custom Annotations](https://mydevgeek.com/spring-boot-rest-web-service-part-3-exception-handling-validation-using-controlleradvice-valid-custom-annotations/)
- When sign the user up, it only give user ```ROLE_USER``` permission. In order to do that, I need to know the Authority object for ```ROLE_USER```, adding ```AuthoritytRepository``` and ```AuthorityService``` code to achieve this.

Front end:
- Design the similar page as login page, and adding the sign up logic.

### CODE REVIEW

File: server/src/main/java/com/bfwg/model/User.java [hunk 1]
```diff
@@ -1,119 +1,128 @@
...
+  @Id
+  @Column(name = "id")
+  @GeneratedValue(strategy = GenerationType.IDENTITY)
+  private Long id;
+
+  @Column(name = "username")
+  private String username;
+
+  // @JsonIgnore
+  @Column(name = "password")
```
Comment (bfwg) [inline server/src/main/java/com/bfwg/model/User.java]: Can we move @JsonIgnore to the password getter? This way, we won't get the password from the JSON response(secure) but we can still set it.

File: server/src/main/java/com/bfwg/rest/UserController.java [hunk 2]
```diff
@@ -1,62 +1,86 @@
...
+
+
+  @RequestMapping(method = POST, value = "/signup")
+  public ResponseEntity<?> addUser(@RequestBody User user, UriComponentsBuilder ucBuilder) {
+
+    User existUser = this.userService.findByUsername(user.getUsername());
+    if (existUser != null) {
+      throw new ResourceConflictException(user.getId(), "Username already exists");
     }
+    List<Authority> auths = this.authService.findByname("ROLE_USER");
```
Comment (bfwg) [inline server/src/main/java/com/bfwg/rest/UserController.java]: Can we relocate the below code into a method in a server? Possibly the user service?

File: server/src/main/java/com/bfwg/model/User.java (hunk 1 above)
Comment (zhengye1) [reply server/src/main/java/com/bfwg/model/User.java]: OK, I can try

File: server/src/main/java/com/bfwg/rest/UserController.java (hunk 2 above)
Comment (zhengye1) [reply server/src/main/java/com/bfwg/rest/UserController.java]: OK, will try tomorrow

File: server/src/main/java/com/bfwg/model/User.java (hunk 1 above)
Comment (zhengye1) [reply server/src/main/java/com/bfwg/model/User.java]: Tried.. but seems like spring boot cannot get the password that user enter in the angular front end with jsonignore..

**Comment by bfwg** [comment]
LGTM 👍

**Pull request merged**